### PR TITLE
combsimp for work on factorials

### PR DESCRIPTION
**Pull request by gxyd**
Now

```
>>> combsimp(factorial(n)*(n + 1))
factorial(n + 1)
```

Fixes issue #9699

### CODE REVIEW

**Comment by leosartaj** [comment]
I think, in case the expr has ``gamma`` expr should be returned with ``gamma``.

In this branch,
```
>>>  e = (x + 1)*factorial(x)/gamma(y)
>>> combsimp(e)
factorial(x + 1)/factorial(y - 1)
```

I think this will be better
```
>>> combsimp(e)
gamma(x + 2)/gamma(y)
```

**Comment by gxyd** [comment]
Yes probably as i think, probably you are right. My reason being that: converting `gamma(y)` to `factorial(y + 1)` would lead to a problem that `y` for `non-integers` the `factorial(y)` when substituted would be returned un-evaluated. But `gamma(y)` would be evaluated.

Can you please tell me why you think so ?

**Comment by leosartaj** [comment]
``gamma`` is defined for non-integers while technically factorial is only defined for positive integers. So, in case the expression has ``gamma`` terms, then I think it's not ideal to convert them to ``factorial``.

**Comment by gxyd** [comment]
@leosartaj @jksuom

**Comment by leosartaj** [comment]
Looks good to me. +1

**Comment by jksuom** [comment]
I think this is ok, but perhaps the docstring should be edited. It does not describe the changed procedure.

**Comment by gxyd** [comment]
@jksuom Do you think [this line](https://github.com/sympy/sympy/blob/master/sympy/simplify/combsimp.py#L33-L34) should be removed ? 
```
If the initial expression contained any
    combinatorial functions, the result is expressed using binomial
```
I think so because for `combsimp(factorial(n))` contains `combinatorial` functions but the final expression `factorial(n)` is not expressed using the `binomial` or `gamma` functions itself.

**Comment by jksuom** [comment]
BTW, this does not solve all problems concerning ``factorial``, since it is reintroduced on [this line](https://github.com/sympy/sympy/blob/master/sympy/simplify/combsimp.py#L86). It seems to be the cause for e.g. this behaviour:
```
In [5]: combsimp(rf(x + n, k)*binomial(n, k))
Out[5]: 
        ⎛n⎞             
(n + x)⋅⎜ ⎟⋅(k + n + x)!
        ⎝k⎠             
────────────────────────
  (k + n + x)⋅(n + x)! 
```

**Comment by gxyd** [comment]
@jksuom  Would it be good: If i put the code introduced in the PR into a function and use it, where the conversion to factorial is there ?

**Comment by jksuom** [comment]
> Do you think this line should be removed ? 

I think yes. Binomials may have originally been planned to have a more central role, but now it seems that they are not very useful in the simplification.

In fact, it looks like the whole process should be reorganized. Probably everything should be first rewritten in terms of the gamma-function. Perhaps it could even be left in that form, at least for now. All this could be the subject of another PR.

> Would it be good: If i put the code introduced in the PR into a function and use it, where the conversion to factorial is there ?

I'm afraid that I cannot quite get your idea here. Perhaps you could elaborate?

**Comment by gxyd** [comment]
Sorry for being late on the issue.
> BTW, this does not solve all problems concerning factorial, since it is reintroduced on this line. It seems to be the cause for e.g. this behaviour:
In [5]: combsimp(rf(x + n, k)*binomial(n, k))
Out[5]: 
        ⎛n⎞             
(n + x)⋅⎜ ⎟⋅(k + n + x)!
        ⎝k⎠             
────────────────────────
  (k + n + x)⋅(n + x)! 

@jksuom  can you please tell me what output do you think be good
```
>>> combsimp(rf(x + n, k)*binomial(n, k))
binomial(n, k)*gamma(k + n + x)/gamma(n + x)
```
Or
```
>>> combsimp(rf(x + n, k)*binomial(n, k))
factorial(n)*factorial(k + n + x - 1)/(factorial(k)*factorial(-k + n)*factorial(n + x - 1))
```
I think that the second is better. Since `rf` being a type of `factorial` ?

**Comment by jksuom** [comment]
This is from the docstring of ``combsimp`` (verbatim).
> The result is be given in terms of binomials and factorials.

It would probably mean the following expression.
```
binomial(n, k)*factorial(k + n + x - 1)/factorial(n + x - 1)
```
It would then be possible to obtain the last form by ``rewrite(factorial)``.

However, the docstring of ``factorial`` says:
> Implementation of factorial function over nonnegative integers.

If we want to follow this (and I think we probably should), the we should use ``gamma`` instead of ``factorial`` unless it is clear that ``x`` would be an integer (in addition to ``n`` and ``k``). This would imply that we use the first one of your two alternatives. Then ``rewrite(factorial)`` would again give the second form, if needed.

I think that ``rf`` is related to ``gamma`` more closely than to ``factorial``. In the definition of ``rf(x, k)`` it is assumed that ``k`` is an integer while ``x`` may have any complex value. So ``rf`` can be expressed in terms of ``factorial`` only for special values of its arguments.

**Comment by gxyd** [comment]
@jksuom Please have a look.

**Comment by jksuom** [comment]
Looks good to me. @aktech ?

File: sympy/simplify/combsimp.py [hunk 1]
```diff
@@ -30,11 +29,11 @@ def combsimp(expr):
     the resulting rising factorial to cancel. Rising factorials with
     the second argument being an integer are expanded into polynomial
     forms and finally all other rising factorial are rewritten in terms
-    more familiar functions. If the initial expression contained any
-    combinatorial functions, the result is expressed using binomial
-    coefficients and gamma functions. If the initial expression consisted
-    of gamma functions alone, the result is expressed in terms of gamma
-    functions.
+    more familiar functions. If the initial expression consisted of
```
Comment (aktech) [inline sympy/simplify/combsimp.py]: `of` more familiar

**Comment by aktech** [comment]
Looks fine. +1

**Comment by gxyd** [comment]
Thanks @jksuom @aktech, I have `rebased` it into a single commit(for cleaner history).

**Comment by jksuom** [comment]
Thanks. I'll merge in a few hours.

**Pull request merged**